basedtyping/__init__.py

```diff
@@ -106,26 +111,35 @@ def __rand__(self, other: object) -> object:
         return Intersection[other, self]
 
 
-if TYPE_CHECKING:
-    Function = Callable[..., object]  # type: ignore[no-any-explicit]
-    """Any ``Callable``. useful when using mypy with ``disallow-any-explicit``
-    due to https://github.com/python/mypy/issues/9496
-
-    Cannot actually be called unless it's narrowed, so it should only really be used as
-    a bound in a ``TypeVar``.
-    """
-else:
-    # for isinstance checks
-    Function = Callable
-
 # Unlike the generics in other modules, these are meant to be imported to save you
 #  from the boilerplate
 T = TypeVar("T")
 in_T = TypeVar("in_T", contravariant=True)
 out_T = TypeVar("out_T", covariant=True)
 Ts = TypeVarTuple("Ts")
 P = ParamSpec("P")
-Fn = TypeVar("Fn", bound=Function)
+
+AnyCallable = Callable[..., object]  # type: ignore[no-any-explicit]
+"""Any ``Callable``. useful when using mypy with ``disallow-any-explicit``
+due to https://github.com/python/mypy/issues/9496
+
+Cannot actually be called unless it's narrowed, so it should only really be used as
+a bound in a ``TypeVar``.
+"""
+
+
+if not BASEDMYPY_TYPE_CHECKING and TYPE_CHECKING:
+    FunctionType: TypeAlias = Callable[P, T]
+else:
+    # TODO: BasedSpecialGenericAlias  # noqa: TD003
+    FunctionType: _SpecialForm = typing._CallableType(types.FunctionType, 2)  # type: ignore[attr-defined]
+
+AnyFunction = FunctionType[..., object]  # type: ignore[no-any-explicit]
+
+TCallable = TypeVar("TCallable", bound=AnyCallable)
+TFunction = TypeVar("TFunction", bound=AnyFunction)
```

Comment:
can we conditionally define these only if python version is < 3.12? since they shouldnt be used unless targeting an old python version

Reply:
lets do some deprecations in another pr